include/ginkgo/core/base/lin_op.hpp

```diff
@@ -855,49 +855,75 @@ public:                                                                      \
                   "semi-colon warnings")
 
 
-#if !(defined(__CUDACC__) || defined(__HIPCC__))
 /**
  * Creates a factory parameter in the factory parameters structure.
  *
  * @param _name  name of the parameter
+ * @param _type  type of the parameter
  * @param __VA_ARGS__  default value of the parameter
  *
  * @see GKO_ENABLE_LIN_OP_FACTORY for more details, and usage example
  *
  * @ingroup LinOp
  */
-#define GKO_FACTORY_PARAMETER(_name, ...)                                    \
-    mutable _name{__VA_ARGS__};                                              \
+#define GKO_FACTORY_PARAMETER(_name, _type, ...)                             \
+    _type mutable _name{__VA_ARGS__};                                        \
                                                                              \
-    template <typename... Args>                                              \
-    auto with_##_name(Args &&... _value)                                     \
+    auto with_##_name(_type _value)                                          \
         const->const ::gko::xstd::decay_t<decltype(*this)> &                 \
     {                                                                        \
-        using type = decltype(this->_name);                                  \
-        this->_name = type{std::forward<Args>(_value)...};                   \
+        this->_name = std::move(_value);                                     \
         return *this;                                                        \
     }                                                                        \
     static_assert(true,                                                      \
                   "This assert is used to counter the false positive extra " \
                   "semi-colon warnings")
-#else  // defined(__CUDACC__) || defined(__HIPCC__)
-// A workaround for the NVCC compiler - parameter pack expansion does not work
-// properly. You won't be able to use factories in code compiled with NVCC, but
-// at least this won't trigger a compiler error as soon as a header using it is
-// included. To not get a linker error, we provide a dummy body.
-#define GKO_FACTORY_PARAMETER(_name, ...)                                    \
-    mutable _name{__VA_ARGS__};                                              \
+
+
+/**
+ * Creates a factory parameter in the factory parameters structure that is a
+ * vector of multiple elements (at most 3).
```

Comment:
Maybe we should actually put back the recursive templated version with non-CUDA so that there is no change there and it works with any number of criteria. For CUDA, we can keep your new version so that it also works there.